### PR TITLE
mam: use mam_pk_t, casts and missing returns

### PR DESCRIPTION
**Pull request by thibault-martinez**

### CODE REVIEW

File: mam/api/api.c [hunk 1]
```diff
@@ -338,9 +338,10 @@ retcode_t mam_api_create_channel(mam_api_t *const api, size_t const height,
     return ret;
   }
 
-  ERR_BIND_RETURN(mam_pk_t_set_add(&api->trusted_channel_ids,
-                                   trits_begin(mam_channel_id(&channel))),
-                  ret);
+  ERR_BIND_RETURN(
+      mam_pk_t_set_add(&api->trusted_channel_ids,
+                       (mam_pk_t *)trits_begin(mam_channel_id(&channel))),
```
Comment (tsvisabo) [inline mam/api/api.c]: this works, but it's also wrong,
`mam_pk_t`  has a `pk` field of `trit_t [MAM_CHANNEL_ID_SIZE]` type, should populate a local `mam_pk_t` and add that to the set

File: mam/api/api.c [hunk 2]
```diff
@@ -396,9 +397,10 @@ retcode_t mam_api_create_endpoint(mam_api_t *const api, size_t const height,
     mam_endpoint_destroy(&endpoint);
     return ret;
   }
-  ERR_BIND_RETURN(mam_pk_t_set_add(&api->trusted_endpoint_ids,
-                                   trits_begin(mam_endpoint_id(&endpoint))),
-                  ret);
+  ERR_BIND_RETURN(
+      mam_pk_t_set_add(&api->trusted_endpoint_ids,
```
Comment (tsvisabo) [inline mam/api/api.c]: ditto

File: mam/mam/message.c [hunk 3]
```diff
@@ -737,7 +737,7 @@ retcode_t mam_msg_read_header(mam_msg_read_context_t *const ctx,
         ret);
     ERR_GUARD_RETURN(0 == ver, RC_MAM_VERSION_NOT_SUPPORTED);
 
-    if (!mam_pk_t_set_contains(trusted_channels_pks, &chid)) {
+    if (!mam_pk_t_set_contains(trusted_channels_pks, (mam_pk_t *)chid)) {
```
Comment (tsvisabo) [inline mam/mam/message.c]: same, `mam_pk_t` is not `trit_t []`

File: mam/mam/message.c [hunk 4]
```diff
@@ -757,7 +757,7 @@ retcode_t mam_msg_read_header(mam_msg_read_context_t *const ctx,
               &spongos_mss, &spongos_wots,
               trits_from_rep(MAM_CHANNEL_ID_SIZE, chid)),
           ret);
-      mam_pk_t_set_add(trusted_channels_pks, ctx->pk);
+      mam_pk_t_set_add(trusted_channels_pks, (mam_pk_t *)ctx->pk);
```
Comment (tsvisabo) [inline mam/mam/message.c]: ditto

File: mam/mam/message.c [hunk 5]
```diff
@@ -766,14 +766,14 @@ retcode_t mam_msg_read_header(mam_msg_read_context_t *const ctx,
               &spongos_mss, &spongos_wots,
               trits_from_rep(MAM_CHANNEL_ID_SIZE, chid)),
           ret);
-      mam_pk_t_set_add(trusted_endpoints_pks, ctx->pk);
+      mam_pk_t_set_add(trusted_endpoints_pks, (mam_pk_t *)ctx->pk);
```
Comment (tsvisabo) [inline mam/mam/message.c]: ditto

File: mam/mam/message.c [hunk 6]
```diff
@@ -766,14 +766,14 @@ retcode_t mam_msg_read_header(mam_msg_read_context_t *const ctx,
...
     } else if (mam_msg_pubkey_epid ==
                pubkey) { /*  absorb tryte epid[81] = 1; */
       ERR_BIND_RETURN(
           mam_msg_unwrap_pubkey_epid(
               &ctx->spongos, msg, trits_from_rep(MAM_CHANNEL_ID_SIZE, ctx->pk)),
           ret);
-      if (!mam_pk_t_set_contains(trusted_endpoints_pks, &ctx->pk)) {
+      if (!mam_pk_t_set_contains(trusted_endpoints_pks, (mam_pk_t *)ctx->pk)) {
```
Comment (tsvisabo) [inline mam/mam/message.c]: ditto

File: mam/mam/tests/test_message.c [hunk 7]
```diff
@@ -138,10 +138,12 @@ static void message_test_generic_read_msg(
   TEST_ASSERT(mam_psk_t_set_add(&psks, pre_shared_key) == RC_OK);
   TEST_ASSERT(mam_ntru_sk_t_set_add(&ntru_sks, ntru) == RC_OK);
   TEST_ASSERT(mam_pk_t_set_add(&trusted_channel_ids,
-                               trits_begin(mam_channel_id(cha))) == RC_OK);
+                               (mam_pk_t *)trits_begin(mam_channel_id(cha))) ==
```
Comment (tsvisabo) [inline mam/mam/tests/test_message.c]: ditto

File: mam/mam/tests/test_message.c [hunk 8]
```diff
@@ -138,10 +138,12 @@ static void message_test_generic_read_msg(
...
+              RC_OK);
   if (ep) {
-    TEST_ASSERT(mam_pk_t_set_add(&trusted_endpoint_ids,
-                                 trits_begin(mam_endpoint_id(ep))) == RC_OK);
+    TEST_ASSERT(mam_pk_t_set_add(
+                    &trusted_endpoint_ids,
+                    (mam_pk_t *)trits_begin(mam_endpoint_id(ep))) == RC_OK);
```
Comment (tsvisabo) [inline mam/mam/tests/test_message.c]: ditto

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: mam/api/api.c (hunk 1 above)
Comment (thibault-martinez) [reply mam/api/api.c]: I initially used a temporary set like you said but ended up just adding a cast to save memory.
I'll add the sets back.

Comment (tsvisabo) [reply mam/api/api.c]: tmp struct `mam_pk_t`?

Comment (thibault-martinez) [reply mam/api/api.c]: Yes

Comment (thibault-martinez) [reply mam/api/api.c]: Oops, said set, sorry

**Review by tsvisabo — DISMISSED** [review]

**Review by tsvisabo — APPROVED** [review]

**Pull request merged**